### ISSUE TITLE
move NetLogo to new Vector implementation in 2.13.2?

### ISSUE THREAD

**Issue by SethTisue** [NetLogo/NetLogo#1823]
https://github.com/scala/scala/pull/8534 will likely land soon and be included in Scala 2.13.2

NetLogo's `LogoList` is just a wrapper around `scala.collection.immutable.Vector[AnyRef]`, so the performance differences should carry over directly to NetLogo

some of the improvements give huge speedups on very long lists. but perhaps in NetLogo such huge lists aren't that common

but also, another change in the new Vector is that small sizes are special-cased. I would expect even ordinary NetLogo models to benefit from that — if they use lists at all, I mean. they can expect to benefit in speed, but at least as important, they will use benefit in lower memory usage. users running with large number of agents, where each agent uses small lists, could see a noticeable drop in RAM usage

and it would be interesting to run the NetLogo benchmark suite before-and-after this change

but.... huge caveat.... you'd have to move to Scala 2.13 first. curious, is any work on that planned or in progress?

**Comment by SethTisue** [comment]
Scala 2.13.2 is now available https://github.com/scala/scala/releases/tag/v2.13.2

**Comment by SethTisue** [comment]
Scala 2.13.3 is out, so the new `Vector` is now pretty well battle-tested

**Comment by EwoutH** [comment]
@SethTisue Did you already try running [scala-collection-compat](https://github.com/scala/scala-collection-compat) over the repository? If so, what where the results?

**Comment by SethTisue** [comment]
@EwoutH I think you might be misunderstanding the purpose of scala-collection-compat. it's useful for using certain 2.13 features from 2.12, but we aren't staying on 2.12 and we don't need those features. I don't see that scala-collection-compat is useful for upgrading to 2.13

**Comment by brandesNW** [comment]
Hi Seth,
What steps need to be taken to upgrade to scala 2.13.2 ?
What version of scala do you think we should aim for (if we have the resources) for the next release (or the one after).
- Aaron
--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Seth Tisue ***@***.***>
Date: Thursday, November 17, 2022 at 10:14 AM
To: NetLogo/NetLogo ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [NetLogo/NetLogo] move NetLogo to new Vector implementation in 2.13.2? (#1823)

@EwoutH<https://urldefense.com/v3/__https:/github.com/EwoutH__;!!Dq0X2DkFhyF93HkjWTBQKhk!RU_A6fhz0RVicRU2hYjmx7jy3tpGZ877uR-NV63_5AJ4r9-XndkuKZb1-C22v9RghQxLXywDOk2lptGm85XsETUdSFCckQm7TA$> I think you might be misunderstanding the purpose of scala-collection-compat. it's useful for using certain 2.13 features from 2.12, but we aren't staying on 2.12 and we don't need those features. I don't see that scala-collection-compat is useful for upgrading to 2.13

**Comment by EwoutH** [comment]
> it's useful for using certain 2.13 features from 2.12, but we aren't staying on 2.12 and we don't need those features.

In that case I think I indeed misunderstood, I thought `Collection213Upgrade` could help upgrade to 2.13 without keeping 2.12 compatiblity.

> ### Collection213Upgrade
> 
> The `Collection213Upgrade` rewrite upgrades to the 2.13 collections without the ability to compile the code-base with 2.12 or 2.11. This rewrite is suitable for applications that don't need to cross-compile against multiple Scala versions.

**Comment by SethTisue** [comment]
Oh, sorry the rewrite rules, I forgot about those. Yes, that could be helpful here. (I thought you meant the scala-collection-compat library.)

**Comment by SethTisue** [comment]
> What version of scala do you think we should aim for (if we have the resources) for the next release (or the one after).

You should go straight to 2.13.10. There's no reason to use any intermediate version as a way station.

You can worry about Scala 3 later. It's best to get on 2.13 first.

> What steps need to be taken to upgrade to scala 2.13.2 ?

Bump the version number, see what breaks, and fix it :-)

You'll want to read https://github.com/scala/scala/releases/v2.13.0 very carefully and absorb it thoroughly.

Anything in the notes in is a potential concern, but by far the biggest area of change is the Scala collections library.

As the notes say, "Most ordinary code that used the old collections will continue to work as-is", but in any large codebase like NetLogo's, "most" will leave you with some work to do.

After bumping `scalaVersion`, you could start by getting  `Test/update` to work — bumping dependency versions if needed (hopefully the recent upgrades have already taken care of most or all of them), and finding alternatives for any dependencies that don't exist for 2.13 (if any — knock on wood!).

Then get the code to compile on 2.13. as suggested by @EwoutH, `Collection213Upgrade` could be helpful here.

Then get the tests to pass.

**Comment by SethTisue** [comment]
I don't know how many warnings the NetLogo codebase generates or whether y'all are in the habit of ignoring them? During the upgrade, there will probably be new warnings that will alert you of potential issues and you won't want to tune them out.

**Comment by SethTisue** [comment]
Oh, and I forgot to mention:

> Upgrading from 2.12? Enable -Xmigration while upgrading to request migration advice from the compiler

(This is in the 2.13.0 release notes, but it's easy to miss.)

**Comment by EwoutH** [comment]
I did try to apply the [Collection213Upgrade](https://github.com/scala/scala-collection-compat#collection213upgrade) rewrite rule using [Scala Steward](https://github.com/scala-steward-org/scala-steward). This rule was [recently added](https://github.com/scala-steward-org/scala-steward/issues/2770) to Scala Steward.

Unfortunately it failed, [here is the log](https://github.com/EwoutH/NetLogo/actions/runs/3752758458/jobs/6375284107#step:6:213).

**Comment by TheBizzle** [comment]
Fixed in PR #2394.

**Comment by EwoutH** [comment]
Wow this is really awesome! Thanks a lot!

Are any benchmarks done already?

**Comment by TheBizzle** [comment]
No, we simply updated to Scala 2.13, but we haven't had a chance to look into how much the upgrade has affected performance.
